### ISSUE TITLE
Some questions regarding the code

### ISSUE THREAD

**Issue by haichaoyu** [csyhhu/MetaQuant#2]
Hi Shangyu,

Great work! I have some questions regarding the code as below
- new_meta_hidden_state_dict is empty [here](https://github.com/csyhhu/MetaQuant/blob/master/meta_utils/helpers.py#L13)?
- meta_opt_flag is unused in [here](https://github.com/csyhhu/MetaQuant/blob/master/meta-quantize.py#L162)?
- what is the setup for training net only after meta_net is well trained? Is there any script? 

Looking forward to your reply.

Thanks

**Comment by csyhhu** [comment]
Hi @haichaoyu ,

Thanks for using my code and finding out the problems !
1. This is a bug in new_meta_hidden_state_dict as you mentioned, I have fixed it. Please refer to [here](https://github.com/csyhhu/MetaQuant/blob/83e7adacfb435234841dccf909f26601fa98ad35/meta_utils/helpers.py#L73) for the change.
2. meta_opt_flag is never used, I have commented it.
3. It is necessary to have meta net cooperatively trained with the base quantized net. My method can not perform training quantized net with a well-trained meta net.

Best regards,
Shangyu

**Comment by haichaoyu** [comment]
Hi @csyhhu,

Encountered another error [here](https://github.com/csyhhu/MetaQuant/blob/master/meta-quantize.py#L232). `layer.calibration` does not exists. Maybe because `forward` in `net` is never called?

**Comment by csyhhu** [comment]
Hi @haichaoyu ,

Sorry for the bug. I have fixed it. Please refresh the project.

Best regards,
Shangyu

**Comment by haichaoyu** [comment]
Hi Shangyu,

Thanks for your quick response. The code now runs successfully. Two more questions to bother you:
- Considering LSTMFC meta_optimizer, in your code, hidden states are detached [here](https://github.com/csyhhu/MetaQuant/blob/master/meta_utils/helpers.py#L73) and meta_net's output gradients are detached [here](https://github.com/csyhhu/MetaQuant/blob/master/meta-quantize.py#L240). So losses will not back-propagate through hidden states and will back-propagate through output only once by [here](https://github.com/csyhhu/MetaQuant/blob/master/meta_utils/meta_quantized_module.py#L88). This way, LSTM actually cannot learn the so-called long-short memory at all. Does this influence the training?
- [Here](https://github.com/csyhhu/MetaQuant/blob/master/meta_utils/meta_quantized_module.py#L88), `(self.calibrated_grads + (self.weight.grad.data - self.calibrated_grads.data).detach())` (term this Equ 1) is used as update to the weight, but during the **real** update [here](https://github.com/csyhhu/MetaQuant/blob/master/meta-quantize.py#L240), only `calibrated_grads` is used instead, which is not the one that is used for back-propagation in Equ 1. Is it intentionally designed?

Best,
Haichao

**Comment by csyhhu** [comment]
Hi @haichaoyu ,

Thanks again for your using.

- For the first question, indeed only the output is used for backpropagation. Hidden state is used only for inference in next iteration's gradient generation. The reason behind it is that if hidden state is incorporated into backpropagation, a (similar) ''max sequence length'' need to be set to avoid memory explode, since I regard each training iteration as a "token" in LSTM treating.  It may be hard to learn the long-short memory. 

- Actually **real** update happen after attaining the refined gradient in [here](https://github.com/csyhhu/MetaQuant/blob/50694404fa010958aa24c96b4c870d5e9442e849/meta-quantize.py#L244). After `optimizee.get_refine_gradient()`, meta gradient will be processed by optimization algorithm (e.g. Adam), then be used to update weight. As you may notice, it is not exactly the same due to the **time gap**:  `self.weight.grad.data` used in forward comes from last  iteration's `self.calibrated_grads` -> `optimizee.get_refine_gradient()`. But it won't harm the training. As I have discussed the same issue in [here](https://github.com/csyhhu/MetaQuant/issues/1), I will try to fix it in latter.

Best regards,
Shangyu
